src/Form/input-classifiers.js

```diff
@@ -46,37 +53,52 @@ const getLargestMeaningfulContainer = (el, form) => {
 }
 
 /**
- * Tries to infer input type, with checks in decreasing order of reliability
- * @type (el: HTMLInputElement, form: HTMLFormElement, Matcher) => Boolean
+ * Tries to infer input subtype, with checks in decreasing order of reliability
+ * @type (el: HTMLInputElement, form: HTMLFormElement, matchers: []Matcher) => string|undefined
  */
-const checkMatch = (el, form, {selector, matcherFn}) => {
-    if (selector && el.matches(selector)) return true
-
-    if (!matcherFn) return false
-
-    return [getExplicitLabelsText(el), el.id, el.placeholder, getRelatedText(el, form)].some(matcherFn)
+const getSubtypeFromMatchers = (el, form, matchers) => {
```

Comment:
This version is more verbose than the previous, but it allows us to run checks in decreasing order of confidence and increasing order of expensiveness (apart from `placeholder` which is cheaper, but gives less confidence than labels). This new function also lets us cache the DOM reads. The previous version was reading all values from the DOM for every single matcher and in the case of `relatedText` it can be pretty expensive. Now we only ever get there if the other checks fail and we read only once per function run.